skimage/segmentation/trainable_segmentation.py

```diff
@@ -0,0 +1,180 @@
+        self.features = None
+
+    def compute_features(self, image):
+        if self.features_func is None:
+            self.features_func = multiscale_basic_features
+        self.features = self.features_func(image)
+
+    def fit(self, labels, image=None):
+        """
+        Train classifier using partially labeled (annotated) image.
+
+        Parameters
+        ----------
+        labels : ndarray of ints
+            Labeled array of shape compatible with ``image`` (same shape for a
+            single-channel image). Labels >= 1 correspond to the training set and
+            label 0 to unlabeled pixels to be segmented.
+        image : ndarray
+            Input image, which can be grayscale or multichannel, and must have a
+            number of dimensions compatible with ``self.features_func``.
+        """
+        if self.features is None:
+            self.compute_features(image)
+        output, clf = fit_segmenter(labels, self.features, self.clf)
+        self.segmented_image = output
+
+    def predict(self, image):
+        """
+        Segment new image using trained internal classifier.
+
+        Parameters
+        ----------
+        image : ndarray
+            Input image, which can be grayscale or multichannel, and must have a
+            number of dimensions compatible with ``self.features_func``.
+
+        Raises
+        ------
+        NotFittedError if ``self.clf`` has not been fitted yet (use ``self.fit``).
+        """
+        if self.features_func is None:
+            self.features_func = multiscale_basic_features
+        features = self.features_func(image)
+        return predict_segmenter(features, self.clf)
+
+
+def fit_segmenter(labels, features, clf):
+    """
+    Segmentation using labeled parts of the image and a classifier.
+
+    Parameters
+    ----------
+    labels : ndarray of ints
+        Image of labels. Labels >= 1 correspond to the training set and
+        label 0 to unlabeled pixels to be segmented.
+    features : ndarray
+        Array of features, with the first dimension corresponding to the number
+        of features, and the other dimensions correspond to ``labels.shape``.
+    clf : classifier object
+        classifier object, exposing a ``fit`` and a ``predict`` method as in
+        scikit-learn's API, for example an instance of
+        ``RandomForestClassifier`` or ``LogisticRegression`` classifier.
+
+    Returns
+    -------
+    output : ndarray
+        Labeled array, built from the prediction of the classifier trained on
+        ``labels``.
+    clf : classifier object
+        classifier trained on ``labels``
+
+    Raises
+    ------
+    NotFittedError if ``self.clf`` has not been fitted yet (use ``self.fit``).
+    """
+    training_data = features[:, labels > 0].T
+    training_labels = labels[labels > 0].ravel()
+    clf.fit(training_data, training_labels)
+    data = features[:, labels == 0].T
+    predicted_labels = clf.predict(data)
```

Comment:
see my other comment. How about leaving the prediction out of here and just returning `clf` as an alternative?